mshadow/base.h

```diff
@@ -1102,5 +1103,68 @@ inline size_t mshadow_sizeof(int type) {
   return size;
 }
 
+
+/*!
+ * \param result the result of a * b is stored in result if the pointer is not nullptr
+ * \return true if multiplication of a and b won't overflow
+ * Uses division method
+ */
+template<typename T>
+inline bool mult_not_overflow_binary(T a, T b, T *result = nullptr) {
```

Comment:
Can we use some built-in utility of gcc to check multiplication overflow: https://gcc.gnu.org/onlinedocs/gcc-5.2.0/gcc/Integer-Overflow-Builtins.html?

Reply:
Good point, can be an optimization if we are under gcc once the PR is working, there's some GPU problems I'm not sure if related to CI instability this weeks in GPU or a deeper problem.